synapse/push/clientformat.py

```diff
@@ -44,6 +44,12 @@ def format_push_rules_for_user(
 
         rulearray.append(template_rule)
 
+        pattern_type = template_rule.pop("pattern_type", None)
```

Comment:
Are you completely sure this works?

1. This mutates template_rule in-place. We explicitly take a copy before mutating it below, which suggests that it's not safe to do this.
2. AFAICS "pattern_type" lives in the same place as "pattern", i.e. it is a key of things in `template_rules["conditions"]`. So I don't think this snippet actually does anything??

Reply:
I do know that without that the rule didn't come down the pushrules/ endpoint. If that corrupts some template or so, that I don't know. Rules not only have a pattern in the conditions, content rules, like the username rule, have them in the rule themselves and no conditions.

Reply:
Also, isn't it mutated 2 lines lower as well? I am a bit confused now.

Reply:
`template_rule = _rule_to_template(r)` above gives you a fresh `template_rule` dict, so not sure why it wouldn't be safe to mutate.

Reply:
looks valid to me.

Reply:
(to confirm that content rules don't have the `pattern` inside a conditions block, but instead have them in the top level: https://spec.matrix.org/v1.4/client-server-api/#default-content-rules. I find it weird. I see why people want to re-write push rules...)

Reply:
> `template_rule = _rule_to_template(r)` above gives you a fresh `template_rule` dict, so not sure why it wouldn't be safe to mutate.

True. I think it was line -58 that had me worried, but perhaps that deepcopy isn't necessary?

Reply:
> Also, isn't it mutated 2 lines lower as well? I am a bit confused now.

Oh you're quite right. I have no idea what that deepcopy is for then.